sematic/db/migrate.py

```diff
@@ -63,26 +80,56 @@ def migrate():
 
         logging.info("Applying {}".format(migration_file))
 
-        with open(
-            os.path.join(get_config().migrations_dir, migration_file), "r"
-        ) as file:
-            sql = file.read()
+        extension = migration_file.split(".")[-1]
+
+        if extension == "sql":
+            _run_sql_migration(migration_file, version)
+
+        if extension == "py":
+            _run_py_migration(migration_file, version)
+
+
+def _run_sql_migration(migration_file: str, version: str):
+    with open(os.path.join(get_config().migrations_dir, migration_file), "r") as file:
+        sql = file.read()
+
+    up_sql = sql.split("-- migrate:down")[0].split("-- migrate:up")[1].strip()
+
+    statements = up_sql.split(";")
 
-        up_sql = sql.split("-- migrate:down")[0].split("-- migrate:up")[1].strip()
+    with db().get_engine().begin() as conn:
+        for statement in statements:
+            if len(statement) == 0:
+                continue
 
-        statements = up_sql.split(";")
+            conn.execute("{};".format(statement))
+
+        _insert_version(conn, version)
+
+
+def _run_py_migration(migration_file: str, version: str):
+    module = importlib.import_module(
+        f"sematic.db.migrations.{migration_file.split('.')[0]}"
+    )
+
+    up, down = getattr(module, "up"), getattr(module, "down")
+
+    try:
+        up()
 
         with db().get_engine().begin() as conn:
-            for statement in statements:
-                if len(statement) == 0:
-                    continue
+            _insert_version(conn, version)
 
-                conn.execute("{};".format(statement))
+    except Exception as e:
+        down()
+        raise e
 
-            conn.execute(
-                text("INSERT INTO schema_migrations(version) values (:version)"),
-                version=version,
-            )
+
+def _insert_version(conn: Connection, version: str):
```

Comment:
This will probably break DBmate. Can we switch to using `migrate.py` exclusively, instead of DBmate? Which means we will want `migrate.py` to support `up`, `down`, and `status`. Ex: `bazel run //sematic/db/migrate_bin -- status`

Reply:
You'll also want to update our developer docs accordingly

Reply:
I don't think this will break dbmate. I don't dbmate will break if there are extra versions in the table.

However, I agree that we should probably migrate to using one or the other exclusively. Let's do this is another PR.

Reply:
Is there a way you can double-check the extra versions don't freak it out? I've run into it complaining about not being able to find migration files before. If you can confirm, I'm cool to delay switching fully to our own thing until later.

Reply:
filed #169 to track. I'll approve now under the assumption you'll confirm

Reply:
`status` and `up` worked file:
```
$ ls sematic/db/migrations
20220424062956_create_runs_table.sql			20220527000512_add_edges_table.sql			20220819172555_external_jobs.sql
20220514015440_create_artifacts_table.sql		20220610234559_create_notes_table.sql			20220830211101_persist_resource_reqs.sql
20220519154144_add_description_tags_to_runs.sql		20220714175433_add_nested_future_id_to_runs.sql		20220915165621_resolution_external_jobs.sql
20220521155045_add_source_code_to_runs.sql		20220723010628_create_users_table.sql			20220921212247_add_exception_json_field.sql
20220521155336_add_root_id_to_runs.sql			20220726001230_add_exception_to_runs_table.sql		20220930014400_migrate_exception_data.py
20220522082435_add_type_serialization_to_artifacts.sql	20220816235619_resolutions.sql				__pycache__

$ DATABASE_URL="sqlite3:/${HOME}/.sematic/db.sqlite3" dbmate status
[X] 20220424062956_create_runs_table.sql
[X] 20220514015440_create_artifacts_table.sql
[X] 20220519154144_add_description_tags_to_runs.sql
[X] 20220521155045_add_source_code_to_runs.sql
[X] 20220521155336_add_root_id_to_runs.sql
[X] 20220522082435_add_type_serialization_to_artifacts.sql
[X] 20220527000512_add_edges_table.sql
[X] 20220610234559_create_notes_table.sql
[X] 20220714175433_add_nested_future_id_to_runs.sql
[X] 20220723010628_create_users_table.sql
[X] 20220726001230_add_exception_to_runs_table.sql
[X] 20220816235619_resolutions.sql
[X] 20220819172555_external_jobs.sql
[X] 20220830211101_persist_resource_reqs.sql
[X] 20220915165621_resolution_external_jobs.sql
[X] 20220921212247_add_exception_json_field.sql

Applied: 16
Pending: 0
```
I suspect `down` will break since it may try to find the latest migration based on the last version inserted.

Is that ok @augray ?

Reply:
Unfortunately I think a functional `down` will be necessary. For example, when one of us uses the shared db, then needs to return it to a prior state for somebody else's use. I also had to do it during a bad deploy with one of our customers once (had accidentally deployed a dev version that had some differences in the migration script). Maybe we can just add `down` support to `migrate.py` right now and leave the rest for later?